kube-client/src/client/client_ext.rs

```diff
@@ -385,6 +395,54 @@ impl Client {
         req.extensions_mut().insert("list");
         self.request::<ObjectList<K>>(req).await
     }
+
+    /// Patch provided `Resource` implementing type `K`
+    ///
+    /// ```no_run
+    /// # use k8s_openapi::api::core::v1::Pod;
+    /// # use k8s_openapi::api::core::v1::Service;
+    /// # use kube::client::scope::{Namespace, Subresource};
+    /// # use kube::prelude::*;
+    /// # use kube::api::{PatchParams, Patch};
+    /// # async fn wrapper() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let client: kube::Client = todo!();
+    /// let pod: Pod = client.get("some_pod", &Namespace::from("default")).await?;
+    /// let pp = &PatchParams::apply("controller").force();
+    /// // Perform an apply patch on the resource
+    /// client.patch(pod, pp, &Patch::Apply(&pod), &Default::default()).await?;
+    /// // Perform an apply patch on the resource status
+    /// client.patch(pod, pp, &Patch::Apply(&pod), &Subresource::Status).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn patch<K, P: Serialize + Debug>(
+        &self,
+        resource: &K,
+        pp: &PatchParams,
+        patch: &Patch<P>,
+        subresource: &Subresource,
+    ) -> Result<K>
+    where
+        K: ResourceExt + Serialize + DeserializeOwned + Clone + Debug,
+        <K as Resource>::DynamicType: Default,
+    {
+        let name = resource.meta().name.as_ref().ok_or(Error::NameResolve)?;
```

Comment:
This name assumption actually severely constrains the usability of `patch` to only `Patch::Apply` and the subset of others that happen to be typed to contain the complete metadata.

I.e. it __generally__ excludes [merge patches](https://github.com/kube-rs/kube/blob/f853afd9b25e87cfccdf57920246eb36f317ed9e/examples/pod_api.rs#L57-L67), [strategic merge patches](https://github.com/kube-rs/kube/blob/f853afd9b25e87cfccdf57920246eb36f317ed9e/kube-client/src/lib.rs#L718-L722) and [json patches](https://github.com/kube-rs/kube/blob/f853afd9b25e87cfccdf57920246eb36f317ed9e/kube-runtime/src/finalizer.rs#L140-L152)

Reply:
My idea/end goal here is to simplify some of the pain points, occurring while using `Apply` patch, without introducing a custom DSL, like in https://pkg.go.dev/k8s.io/client-go/applyconfigurations, just to provide a apiVersion and kind for the apply patch. 

It makes sense to address this incompatibility with other patch types. Need to think it through